sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c

```diff
@@ -287,18 +294,131 @@ int _fopts_mlme_link_check_req(lorawan_buffer_t *buf)
+        /* Indicate error */
+        return 0;
+    }
+
+    /* Mark the first contiguous block */
+    mac->mlme.adr_flags |= 0x8;
+    uint8_t c = index;
+    while (c < len) {
+        if (p[c] == GNRC_LORAWAN_CID_LINK_ADR_REQ) {
+            if ((len - c) >= GNRC_LORAWAN_FOPT_LINK_ADR_REQ_SIZE) {
+                /* There's a LinkADRReq */
+                mac->mlme.adr_req_cnt++;
+                consumed_bytes += GNRC_LORAWAN_FOPT_LINK_ADR_REQ_SIZE;
+            }
+            else {
+                /* The current entry overflow */
+                return 0;
+            }
+        }
+        else {
+            break;
+        }
+        /* Update cursor */
+        c = index + consumed_bytes;
+    }
+
+    int req_index = index + consumed_bytes - GNRC_LORAWAN_FOPT_LINK_ADR_REQ_SIZE;
+
+    uint8_t dr = p[req_index + 1] >> 4;
+    uint16_t tx_power = p[req_index + 1] & 0xF;
+    uint16_t channel_mask = (p[req_index + 3] << 8 ) | p[req_index + 2];
+    uint8_t chmask_ctrl = (p[req_index + 4] & 0x70) >> 4;
+    uint8_t nbtrans = p[req_index + 4] & 0xF;
+
+    int status = 0;
+    gnrc_lorawan_set_uncnf_redundancy(mac, ((nbtrans != 0) ? (uint8_t)(nbtrans - 1) \
+                                             : CONFIG_LORAMAC_DEFAULT_REDUNDANCY));
+
+    if (gnrc_lorawan_set_tx_power(mac, tx_power) >= 0) {
+        status |= 0x4;
+    }
+
+    if (gnrc_lorawan_validate_dr(dr) == true) {
+        mac->last_dr = dr;
+        status |= 0x2;
+    }
+
+    if (chmask_ctrl == 6) {
+        channel_mask = 0;
+        for (unsigned i = 0; i < GNRC_LORAWAN_MAX_CHANNELS; i++) {
+            if (mac->channel[i]) {
+                channel_mask |= 1;
+            }
+            channel_mask = channel_mask << 1;
+        }
+        int res = gnrc_lorawan_phy_set_channel_mask(mac, channel_mask);
+        (void) res;
+        assert(res >= 0);
+        status |= 0x1;
+    }
+    else if (chmask_ctrl == 0) {
+        /* Try to apply channel mask */
+        if (gnrc_lorawan_phy_set_channel_mask(mac, channel_mask) >= 0) {
+            status |= 0x1;
+        }
+    }
+
+    /* Use the last 3 bits for the status */
+    mac->mlme.adr_flags |= status;
+
+    /* Set `LinkADRAns` for next uplink */
+    mac->mlme.pending_mlme_opts |=  GNRC_LORAWAN_MLME_OPTS_LINK_ADR_ANS;
+
+    DEBUG("gnrc_lorawan_mlme: LinkADRReq DataRate_TXPower : DR%u TX%u\n", dr, tx_power);
+    DEBUG("gnrc_lorawan_mlme: LinkADRReq ChMaskCntrl : %u and Channel mask : \
+           %u\n", chmask_ctrl, channel_mask);
+    DEBUG("gnrc_lorawan_mlme: LinkADRReq Redundancy : %u\n", nbtrans);
+
+    return consumed_bytes;
+}
+
+int _fopts_mlme_link_adr_ans(gnrc_lorawan_t *mac, lorawan_buffer_t *buf)
+{
+    for (uint8_t _count = mac->mlme.adr_req_cnt; buf && _count; _count--) {
+        assert(buf->index + GNRC_LORAWAN_CID_LINK_ADR_ANS_SIZE <= buf->size);
+        buf->data[buf->index++] = GNRC_LORAWAN_CID_LINK_ADR_ANS;
+        buf->data[buf->index++] = mac->mlme.adr_flags;
```

Comment:
```suggestion
        buf->data[buf->index++] = mac->mlme.adr_flags & GNRC_LORAWAN_ADR_ANS_MASK;
```

Where `GNRC_LORAWAN_ADR_ANS_MASK=0x7`. Note `adr_flags` is keeping track of the contiguous block too, therefore the current snippet will always add an extra bit.